File: Choonghun/개인 문제/6주차/프로그래머스 42628 - 이중우선순위큐/이중우선순위큐.py
def solution(operations):
    que = []
    for op in operations:
        com, num = op.split()
        num = int(num)
        if com == "I": # 삽입
            que.append(num) 
            i = len(que)-1
            while i > 0: # 정렬
                if que[i-1] > que[i]:
                    que[i],que[i-1] = que[i-1], que[i]
                else:
                    break
                i-=1
        else: # pop
            if len(que) > 0:
                if num == 1: # 최댓값 pop
                    que.pop()
                else: # 최솟값 pop
                    que.pop(0)
    
    if len(que) > 0: # 큐에 값이 있으면
        answer = [que[-1], que[0]] #최댓값, 최솟값
    else: # 없으면
        answer = [0,0]
    return answer

File: Gwanghun/개인 문제/9주차/프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.cpp
#include <string>
#include <vector>
#include <queue>
#include <map>

using namespace std;

vector<int> cal(vector<string> operations) {
    
    priority_queue<int> UpPq;
    priority_queue<int, vector<int>, greater<int>> DownPq;
    map<int,int> m;
    
     for(int i=0; i<operations.size(); i++){
         if(operations[i][0]== 'I'){
             string val="";
             for(int j=2; j<operations[i].size(); j++){
                 val+=operations[i][j];
             }
             int value=stoi(val);
             UpPq.push(value);
             DownPq.push(value);
             m[value]++;
         }else{
             if(operations[i][2]== '-'){
                 while(!DownPq.empty()&&m[DownPq.top()]==0){
                     DownPq.pop();
                 }
                 if(!DownPq.empty()){
                     m[DownPq.top()]--;
                     DownPq.pop();
                 }
             }
             else{
                 while(!UpPq.empty() && m[UpPq.top()]==0){
                     UpPq.pop();
                 }
                 if(!UpPq.empty()){ 
                     m[UpPq.top()]--;
                     UpPq.pop();
                 }
             }
         }
     }
    
    while(!UpPq.empty()&&m[UpPq.top()]==0){
        UpPq.pop();
    }

    while(!DownPq.empty()&&m[DownPq.top()]==0){
        DownPq.pop();
    }
    
    if(DownPq.empty()||UpPq.empty()){
        vector<int> answer;
        answer.push_back(0);
        answer.push_back(0);
        return answer;
    }
    
    else {
        vector<int> answer;
        answer.push_back(UpPq.top());
        answer.push_back(DownPq.top());     
        return answer;
    }
}

vector<int> solution(vector<string> operations) {
    vector<int> answer = cal(operations);
    return answer;
}
